crates/hir_ty/src/diagnostics/match_check/deconstruct_pat.rs

```diff
@@ -611,291 +688,334 @@ impl SplitWildcard {
-        }
+    pub(super) fn is_or_pat(&self) -> bool {
+        matches!(self.ctor, Or)
     }
 
-    /// Replaces contained fields with the arguments of the given pattern. Only use on a pattern
-    /// that is compatible with the constructor used to build `self`.
-    /// This is meant to be used on the result of `Fields::wildcards()`. The idea is that
-    /// `wildcards` constructs a list of fields where all entries are wildcards, and the pattern
-    /// provided to this function fills some of the fields with non-wildcards.
-    /// In the following example `Fields::wildcards` would return `[_, _, _, _]`. If we call
-    /// `replace_with_pattern_arguments` on it with the pattern, the result will be `[Some(0), _,
-    /// _, _]`.
-    /// ```rust
-    /// let x: [Option<u8>; 4] = foo();
-    /// match x {
-    ///     [Some(0), ..] => {}
-    /// }
-    /// ```
-    /// This is guaranteed to preserve the number of patterns in `self`.
-    pub(super) fn replace_with_pattern_arguments(
-        &self,
-        pat: PatId,
-        cx: &MatchCheckCtx<'_>,
-    ) -> Self {
-        // FIXME(iDawer): Factor out pattern deep cloning. See discussion:
-        // https://github.com/rust-analyzer/rust-analyzer/pull/8717#discussion_r633086640
-        let mut arena = cx.pattern_arena.borrow_mut();
-        match arena[pat].kind.as_ref() {
-            PatKind::Deref { subpattern } => {
-                assert_eq!(self.len(), 1);
-                let subpattern = subpattern.clone();
-                Fields::from_single_pattern(arena.alloc(subpattern))
+    pub(super) fn ctor(&self) -> &Constructor {
+        &self.ctor
+    }
+
+    pub(super) fn ty(&self) -> &Ty {
+        &self.ty
+    }
+
+    pub(super) fn iter_fields<'a>(&'a self) -> impl Iterator<Item = &'a DeconstructedPat<'a>> + 'a {
+        self.fields.iter_patterns()
+    }
+
+    /// Specialize this pattern with a constructor.
+    /// `other_ctor` can be different from `self.ctor`, but must be covered by it.
+    pub(super) fn specialize<'a>(
+        &'a self,
+        cx: &MatchCheckCtx<'_, 'p>,
+        other_ctor: &Constructor,
+    ) -> SmallVec<[&'p DeconstructedPat<'p>; 2]> {
+        match (&self.ctor, other_ctor) {
+            (Wildcard, _) => {
+                // We return a wildcard for each field of `other_ctor`.
+                Fields::wildcards(cx, &self.ty, other_ctor).iter_patterns().collect()
             }
-            PatKind::Leaf { subpatterns } | PatKind::Variant { subpatterns, .. } => {
-                let subpatterns = subpatterns.clone();
-                let subpatterns = subpatterns
-                    .iter()
-                    .map(|field_pat| (field_pat.field, arena.alloc(field_pat.pattern.clone())));
-                self.replace_with_fieldpats(subpatterns)
+            (Slice(self_slice), Slice(other_slice))
+                if self_slice.arity() != other_slice.arity() =>
+            {
+                unimplemented!()
```

Comment:
Can people hit this currently? We don't wanna panic if that is the case I'd say

Reply:
No, `Slice` is void struct. It is here just to preserve semantics of catchall arms

```rust
/// A constructor for array and slice patterns.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub(super) struct Slice {
    _unimplemented: Void,
}
```

Reply:
To 'preserve' in relation of rustc